### PR TITLE
Add support for base-prefixed integers: '0b', '0o', '0x'

### PR DESCRIPTION
**Pull request by usedbytes**
Added to the language in toml-lang/toml#507

Fixes: #7

### CODE REVIEW

**Comment by arp242** [comment]
I made a change in #274 to disallow leading zeros in numbers (i.e. `i = 001` is specifically disallowed by the TOML specification) as tests were failing on that.

This conflicts with this patch a bit; sorry about that 😅 I think the patch looks mostly okay from a quick review; if you're still interested in working on this you can rebase on master, fix it up if needed, and I'll merge it. If not then that's okay and I'll fix it up myself.

Thanks!

**Comment by usedbytes** [comment]
Gosh I'd totally forgotten I typed this 😆 I'll have a look at rebasing this evening.
I also didn't notice all the other PRs implementing this when I did it! Woops.

**Comment by arp242** [comment]
> Gosh I'd totally forgotten I typed this

Yeah, it's been a while 😅 This project was unmaintained for a while and I took over day-to-day maintainership a few days ago, hence the late response.

No hurries getting it fixed; I'm still working on some of the bugs etc. that were reported. Just figured I'd give you the opportunity in case you were keen.

> I also didn't notice all the other PRs implementing this when I did it! Woops.

No worries, and this PR was the best one so there's that :-)

**Comment by usedbytes** [comment]
Well, your new test cases have shown some bugs in my implementation, and it's also apparently not spec conformant (`-` is not allowed on prefixed-integers) :blush: 

Do you have any preference for how much validation is done in `lex` versus `parse`? For example, on floats and dates, quite a lot of the validation is done in `parse` after `lex` has pulled out something "approximate". For integers, if `ParseInt()` fails, that's treated as a `p.bug()`.

It feels more in-line with the rest of the code if I don't be too strict in `lex`, and let `ParseInt()` fail for some malformed inputs (such as `0o8`), but let me know if you have a preference either way.

**Comment by arp242** [comment]
Only getting 100% valid numbers in lex is rather tricky as you need to look ahead or backtrack quite a bit to be sure. I suppose that this is why some of the number validation was already in parse rather than lex, and this is also why I added the leading zero check in there. I tried doing that in lex and it wasn't all that easy. Just seeing `0` doesn't mean it's a number, but can also be `0007-01-01` for the 1st of January in the year 7.

Getting an as clean as possible output from the lexer is nice, but it's main task is not to fully parse numbers and the like; just split things in to "this is a number (possible invalid, but a number nonetheless)", "this is a string", "this is a key", "this is a date", etc.

> For integers, if ParseInt() fails, that's treated as a p.bug().

I think that's okay, as there are a number of checks for this (`numPeriodsOK`, `numUnderscoresOK`, `numHasLeadingZero`). Any other invalid numbers should probably be detected in that way so a reasonably nice friendly error message can be presented. For example this:

    i, err := strconv.ParseInt("1__000", 0, 64)

Will give the error:

    strconv.ParseInt: parsing "1__000": invalid syntax

Which is kind of meh. "Invalid" in what way? With the check that's here now you get `Invalid integer "1__000": underscores must be surrounded by digits`. Much nicer ya?

If there are any unhandled situations that will result in `ParseInt` failing then that's probably a bug.

Anyway, in short, stuff like `0xzz` and `0b99` can just be detected by the lexer, as that's clearly not valid and not hard to detect. But more tricky stuff: just check for it in the parser. If you find yourself trying to get it done in the lexer and getting stuck then just move it to the parser.

File: parse.go [hunk 1]
```diff
@@ -183,7 +183,7 @@ func (p *parser) value(it item) (interface{}, tomlType) {
 				it.val)
 		}
 		val := strings.Replace(it.val, "_", "", -1)
-		num, err := strconv.ParseInt(val, 10, 64)
+		num, err := strconv.ParseInt(val, 0, 64)
```
Comment (arp242) [inline parse.go:186]: Since you changed the base to `0` you can remove that `strings.Replace` the line above this. `ParseInt` permits underscores if the base is 0 (and *only* of it's 0, for some reason, I think they did that for compatibility when the new literals were added in Go 1.13).

Comment (usedbytes) [reply parse.go]: ack

**Comment by usedbytes** [comment]
OK I've pushed a fairly significant rewrite. It's stricter (and more correct) than the first version, but I have to admit there's more states and special cases than I'd really like, but I didn't figure a better way.

**Comment by arp242** [comment]
Cheers, at a glance it seems mostly good aside from perhaps a few minor things. I'll take a proper look soon this evening or tomorow.

**Review by arp242 — COMMENTED** [review]
Code mostly seems good; I walked a bit through the lexer code, and I can't see any obvious way to do it more concisely, and overall it seems plenty clear enough to me 👍

Just a few minor comments wrt. tests and error messages, nothing major.

**Comment by arp242** [comment]
Alright, I went ahead and fixed the minor comments; the test failures are because the `toml-test` repo changed, and should be solved once it's merged to master – they're not related to these changes.

Thank you for your work 😊

File: parse.go [hunk 2]
```diff
@@ -362,10 +362,15 @@ func numUnderscoresOK(s string) bool {
...
-			continue
 		}
-		accept = true
+
+		// isHexadecimal is a superset of all the permissable characters
+		// surrounding an underscore.
+		if isHexadecimal(r) {
+			accept = true
+		} else {
+			accept = false;
```
[the remark this answers is not in the document]
Comment (usedbytes) [reply parse.go]: 😅oops, too used to writing C

**Comment by usedbytes** [comment]
Nice, thanks! And thanks for maintaining this project 😀

File: parse.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (arp242) [reply parse.go]: > I think a test-case was failing when I did accept = isHexadecimal(r) but I don't remember exactly what I did.

Don't know? It seems to work fine after I changed it. Maybe you mixed up `=` and `:=`? It's a common and easy mistake to make.
